### ISSUE TITLE
False negative with `Class1 | Class2` union syntax prior to Python 3.10, when `__future__.annotations` is imported

### ISSUE THREAD

**Issue by aThorp96** [microsoft/pyright#6776]
**Describe the bug**
When using the Union type syntax for Python 3.10+, e.g. `str | None`, with python version <3.10, pyright incorrectly finds no errors if `__future__.annotations` is imported. 

**Code or Screenshots**
Given the file
```python
def fn(input: str | None):
    pass
```

Pyright correctly finds the bug when Python version is below 3.10:
> `error: Alternative syntax for unions requires Python 3.10 or newer`

However if you import `__future__.annotations` Pyright falsely claims there are no type errors

<img width="1106" alt="Screenshot 2023-12-18 at 4 05 20 PM" src="https://github.com/microsoft/pyright/assets/28596783/0134b7eb-6128-4e95-b922-bf78dadf8d31">

**VS Code extension or command-line**

I am working with the command line version `1.1.341`

**Comment by erictraut** [comment]
This is by design, not a bug. In fact, pyright contains significant amounts of logic to implement the behavior you're seeing.

If you use `|` in a type annotation expression and annotation evaluations are deferred (because of the presence of a `__future__.annotations` import), then the runtime will not attempt to evaluate the `|`, and it can be safely used within the type annotation.

Note that if you attempt to use `|` in a runtime-evaluated expression (outside of a type annotation), pyright will emit an error. This mirrors the behavior you see if you attempt to run this code.

```python
from __future__ import annotations
from typing import cast

x: int | str = 0 # No Error
y = cast(int | str, 0) # Error
```

**Comment by mikenerone** [comment]
@erictraut
> This is by design, not a bug. In fact, pyright contains significant amounts of logic to implement the behavior you're seeing.
I'm sorry, I don't understand. I get how the deferral prevents a `TypeError` at runtime (because it turns the `|` operands into strings), but I don't understand why it's desirable behavior for Pyright not to alert me when I've used a `|` type annotation in a version of Python where that syntax is meaningless. Isn't the message `error: Alternative syntax for unions requires Python 3.10 or newer` just as true for Python 3.9 regardless of the deferred annotations?

**Comment by erictraut** [comment]
The `|` syntax isn't meaningless. It's syntactically correct. It's a semantically-valid expression and has a well-defined meaning to a static type checker. A static type checker has no problem interpreting this (correctly) as a union. The only concern is the runtime, which may be too old to properly interpret the expression if it is evaluated. However, if `from __future__ import annotations` is used or the type annotation expression is enclosed in quotes, then the runtime doesn't evaluate the expression, and it's no problem.

> It seems alarmingly easy to forget you're working in a Python < 3.10 codebase and use a new-style union type annotation expression out of habit, and doubly alarming that my type checker is going to silently ignore that I've done so.

I don't understand why this alarming to you. If a static type checker is able to properly evaluate the type annotation and the runtime doesn't care about the type annotation, why do you want to be alerted to an error?

> can you suggest some other static analyzer that I should have in place to alert me

Alert you of what? A non-existent error condition? I don't know of any static type checkers or linters that would report a problem here because there is no problem to report.

**Comment by mikenerone** [comment]
@erictraut Ah, I failed to understand the implication that the type check actually _works_ with the new syntax, even with Python < 3.10. In our case, we actually run into an error due to a runtime type check, which I felt like should have been caught earlier than that, but I understand your rationale now. It _does_ seem like something that might be an option so that projects doing significant amounts of runtime type checking could have a way to tell Pyright to catch such mistakes during pre-commit checks, but it's admittedly less critical than I'd been thinking.

**Comment by aThorp96** [comment]
I guess I am not sure I follow the reasoning here: 

> It's a semantically-valid expression and has a well-defined meaning to a static type checker. A static type checker has no problem interpreting this (correctly) as a union. The only concern is the runtime, which may be too old to properly interpret the expression if it is evaluated. 

This makes sense, but I would have expected a type checker to inform of syntactically invalid type signatures given a specific python version. I mean, it does: `error: Alternative syntax for unions requires Python 3.10 or newer`. I don't follow why importing `__future__.annotations` would justify not the emitting of that error. Pyright doesn't guarantee that the type signature isn't evaluated at runtime, as evidenced by the below snippet. I think that's okay, it can't and IMO shouldn't guarantee that. Just because a type signature wouldn't be evaluated at runtime doesn't change whether or not that it's valid for the given python version. Avoiding that kind of runtime "doesn't matter the type is wrong so long as I don't step on a landmine" is precisely why I like using pyright in the first place.

I guess my concern is that my language server and static code analyzer is not in alerting me if I'm using invalid syntax for the configured python version, which regardless of runtime anything seems like a bug 

<img width="1113" alt="Screenshot 2023-12-19 at 4 06 01 PM" src="https://github.com/microsoft/pyright/assets/28596783/f8933b6e-ba52-4050-a234-120723ae9bc1">

**Comment by erictraut** [comment]
> I guess my concern is that my language server and static code analyzer is not in alerting me if I'm using invalid syntax for the configured python version

This is valid _syntax_ in all versions of Python. It will be parsed by the Python parser without any syntax errors and will properly convert it into a valid AST. The problem is that some versions of Python generate a runtime exception when _evaluating_ this expression. If the expression is _not evaluated_ by the runtime (e.g. because it's enclosed in quotes or because `__future__.annotations` is imported), then there is no issue from a static type checking standpoint.

If you're using a runtime type checker that introspects type annotation expressions and attempts to evaluate them, then its the responsibility of your runtime type checker to handle any potential errors that it encounters. For example, the runtime type checker could implement its own evaluation logic to properly evaluate the `|` when `eval` raises an exception. From a static typing perspective, these expressions are perfectly valid, have unambiguous semantic meanings, and present no problem. A static type checker has no knowledge of whether you're using a runtime type checker and whether it has limitations based on the version of Python that you're using. Nor should it.

Many pyright users rely on the fact that pyright doesn't consider this an error. As I mentioned above, there's significant logic in place to implement this behavior. This isn't something I'd consider changing.

It would be possible to add a configuration setting to provide the alternate behavior you're requesting, but I'm reluctant to add such a mode without additional strong signal from pyright users. This is the first time (to the best of my recollection) that anyone has ever requested this feature, so it's not something I'm inclined to add at this time.

This problem will also resolve itself over time as the older versions of Python age out, so to the extent that this is a problem today, it will become less so over time.

In the meantime, if you want pyright to treat type annotations as invalid based on whether the configured version of Python can evaluate them at runtime, then my recommendation is: 1) don't use `from __future__ import annotations` in your code base (perhaps use a linter rule to ban its use), and 2) avoid using quoted type annotations except where absolutely required for forward references.

**Comment by mikenerone** [comment]
Just to clarify, it's not a matter of checking an `eval`. Runtime type checking isn't typically working with a string of source code that it needs to eval. The type expressions were already evaluated by the interpreter at import time. Any type checking code just attempts to inspect the resulting typing objects. This is relatively simple. Unfortunately, In the cases we're talking about, the `__future__` import conspires to turn the resulting object into a concatenated string of type names instead. Re-evaluating the source code as you suggest would essentially require reimplementing a full static analyzer intended for runtime use. That would take a lot of time and significantly deeper knowledge to implement correctly. This is logic that you, thankfully (seriously!), have and have already applied, and we'd love to leverage that to solve the problem with this corner case.

**Comment by aThorp96** [comment]
Would you be open a PR to emit a warning instead of an error when the new Union syntax is used on an older python version and is `__future__.annotations` imported? Just so that it's not silent?

**Comment by erictraut** [comment]
> Would you be open a PR to emit a warning instead of an error when the new Union syntax is used on an older python version and is __future__.annotations imported?

No, that wouldn't be consistent with the way diagnostics work in pyright. Diagnostic severities are controlled on a per-rule basis and are user-configurable. We don't hard-code any diagnostics as "warnings", "errors", etc.

As I said, if you're concerned about this, don't use an import of `__future__.annotations` in your code.
